src/Dev/State/FixtureTestState.php

```diff
@@ -0,0 +1,229 @@
+<?php
+
+namespace SilverStripe\Dev\State;
+
+use LogicException;
+use SilverStripe\Control\Director;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Core\Manifest\ClassLoader;
+use SilverStripe\Dev\FixtureFactory;
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Dev\YamlFixture;
+use SilverStripe\ORM\DataObject;
+
+class FixtureTestState implements TestState
+{
+
+    /**
+     * @var FixtureFactory[]
+     */
+    private $fixtureFactories = [];
+
+    /**
+     * Called on setup
+     *
+     * @param SapphireTest $test
+     */
+    public function setUp(SapphireTest $test)
+    {
+        if ($this->testNeedsDB($test)) {
+            $tmpDB = $test::tempDB();
+            if (!$tmpDB->isUsed()) {
+                $tmpDB->build();
+            }
+            DataObject::singleton()->flushCache();
+
+            if (!$tmpDB->hasStarted()) {
+                foreach ($test->getRequireDefaultRecordsFrom() as $className) {
+                    $instance = singleton($className);
+                    if (method_exists($instance, 'requireDefaultRecords')) {
+                        $instance->requireDefaultRecords();
+                    }
+                    if (method_exists($instance, 'augmentDefaultRecords')) {
+                        $instance->augmentDefaultRecords();
+                    }
+                }
+                $this->loadFixtures($test);
+            }
+            $tmpDB->startTransaction();
+        }
+    }
+
+    /**
+     * Called on tear down
+     *
+     * @param SapphireTest $test
+     */
+    public function tearDown(SapphireTest $test)
+    {
+        if ($this->testNeedsDB($test)) {
+            $test::tempDB()->rollbackTransaction();
```

Comment:
Ok, follow up to https://github.com/silverstripe/silverstripe-framework/issues/8182, we need to check if this rollback failed.

If it failed, we need to destroy the database instead. That will cause the next setUp() to run `$tmpDB->build();`. 

Easy fix @dhensby ?

Reply:
You'll need to rebuild fixture factories as well, since those normally only get rebuilt on setuponce.